### PR TITLE
feat: add fill logic

### PR DESCRIPTION
**Pull request by chrismaree**
This PR implements the minimum viable relayer, adding in deposit validation and fill logic.

### CODE REVIEW

File: src/HubPoolEventClient.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+  public firstBlockToSearch: number;
+
+  constructor(
+    readonly hubPool: Contract,
+    readonly startingBlock: number = 0,
+    readonly endingBlock: number | null = null
+  ) {}
+
+  async computeRealizedLpFeePctForDeposit(deposit: Deposit) {
+    return toBNWei(0.1);
```
Comment (chrismaree) [inline src/HubPoolEventClient.ts:19]: this still needs to be done to truly have an MVP relayer.

File: src/HubPoolEventClient.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import { spreadEvent, assign, Contract, toBNWei, BigNumber, toBN } from "./utils";
+import { Deposit, Fill, SpeedUp } from "./interfaces/SpokePool";
+import { destinationChainId } from "../test/utils";
+
+export class HubPoolEventClient {
+  private whitelistedRoutes: {
```
Comment (nicholaspai) [inline src/HubPoolEventClient.ts:6]: FYI I think we're going to remove this `whitelistedRoute` in favor of a `poolRebalanceRoute` in [this PR](https://github.com/across-protocol/contracts-v2/pull/89).

File: src/HubPoolEventClient.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    readonly startingBlock: number = 0,
+    readonly endingBlock: number | null = null
+  ) {}
+
+  async computeRealizedLpFeePctForDeposit(deposit: Deposit) {
+    // TODO: implement this method.
+    return toBNWei(0.1);
+  }
+
+  getDestinationTokenForDeposit(deposit: Deposit) {
```
Comment (nicholaspai) [inline src/HubPoolEventClient.ts:23]: This `destinationToken` will soon be emitted in the SpokePool's `FundsDeposited` event from this [PR](https://github.com/across-protocol/contracts-v2/pull/89).

File: src/HubPoolEventClient.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    return this.whitelistedRoutes;
+  }
+
+  validateFillForDeposit(fill: Fill, deposit: Deposit) {
+    // TODO: this method should validate the realizedLpFeePct and the destinationToken for the fill->deposit relationship.
+    return true;
+  }
+
+  async update() {
+    const searchConfig = [this.firstBlockToSearch, this.endingBlock || (await this.getBlockNumber())];
```
Comment (nicholaspai) [inline src/HubPoolEventClient.ts:37]: Is there any reason to paginate event search on L1?

File: src/Relayer.ts [hunk 5]
```diff
@@ -1,15 +1,55 @@
...
+    const unfilledDeposits = this.getUnfilledDeposits().sort((a, b) =>
+      a.unfilledAmount.mul(a.deposit.relayerFeePct).lt(b.unfilledAmount.mul(b.deposit.relayerFeePct)) ? 1 : -1
+    );
+
+    // Fetch the realizedLpFeePct for each unfilled deposit. Execute this in parallel as each requires an async call.
+    const realizedLpFeePcts = await Promise.all(
+      unfilledDeposits.map((deposit) => this.hubPoolClient.computeRealizedLpFeePctForDeposit(deposit.deposit))
+    );
+
+    if (unfilledDeposits.length > 0)
```
Comment (nicholaspai) [inline src/Relayer.ts:27]: nit: move this closer to the instantiation of `unfilledDeposits`

File: src/Relayer.ts [hunk 6]
```diff
@@ -1,15 +1,55 @@
...
+    else this.logger.debug({ at: "Relayer", message: "No unfilled deposits" });
+    
+    // Iterate over all unfilled deposits. For each unfilled deposit add a fillRelay tx to the multicallBundler.
+    for (const [index, unfilledDeposit] of unfilledDeposits.entries()) {
+      const destinationToken = this.hubPoolClient.getDestinationTokenForDeposit(unfilledDeposit.deposit);
+      this.multicallBundler.addTransaction(this.fillRelay(unfilledDeposit, destinationToken, realizedLpFeePcts[index]));
+    }
+  }
+
+  // TODO: right now this method will fill the whole amount of the relay. Next iteration should consider the wallet balance.
```
Comment (nicholaspai) [inline src/Relayer.ts:38]: Consider adding a placeholder for a profitability module to check that a relay is profitable

**Review by nicholaspai — APPROVED** [review]
I like the structure +1. It seems that adding `destinationToken` to the `FundsDeposited` event is going to simplify this code, what do you think?

File: src/HubPoolEventClient.ts (hunk 2 above)
Comment (chrismaree) [reply src/HubPoolEventClient.ts:6]: that's ok. will re-release the contrtacts when that's done and update accordinaly.

File: src/HubPoolEventClient.ts (hunk 4 above)
Comment (chrismaree) [reply src/HubPoolEventClient.ts:37]: this is not pagination. it is to ensure we don't search over the same event period twice. i.e if run one goes from block 0->10 the next run should go from 11->current block.

File: src/Relayer.ts (hunk 6 above)
Comment (chrismaree) [reply src/Relayer.ts:38]: sure.

**Pull request merged**